mesg.1
.\"	@(#)mesg.1	2.41
.\"
.TH MESG 1 "Jan 20, 2000"
.AT 3
.SH NAME
mesg \- permit or deny messages
.SH SYNOPSIS
.B mesg
[-s] [-v] [y|n|ye|ne|Y|N|NE] [d] [-p[w|t|k|a]] [-x[w|t|k|n]] [-m[l|c|a]] [-h[Y|y|n]] [-r[y|n]] [-b[y|n]]
.SH DESCRIPTION
This is the "Orville write" verison of the standard Unix
.I mesg
command.
.PP
.I Mesg
with argument
.B n
forbids messages via
.IR  write (1),
.IR jot (1),
.IR tel (1),
and
.IR talk (1)
by revoking non-user
write permission on the user's terminal.
.I Mesg
with argument
.B y
reinstates permission.
All by itself,
.I mesg
reports the current state without changing it.
.PP
The
.B ne
and
.B ye
settings mean ``no with exceptions'' and ``yes with exceptions'' respectively.
If
.B ne
is set, and there is file named
.I .yeswrite
in your home directory,
then the users whose logins are listed there may still write you.
If
.B ye
is set, and there is a file named
.I .nowrite
in your home directory,
then the users whose logins are listed there may not write you.
These files have no effect if the permissions are set to
.B n
or
.BR y .
The
.I .nowrite
and
.I .yeswrite
files do not need to be permitted to anyone else,
and almost any plausible format will be understood
(listing one login name per line is a good default).
Lines may be commented out with a # sign in the first column.
.PP
The upper case
.B Y
and
.B N
do all that the lower case ones do, but may have some additional affects
depending on the installation.
.PP
The
.B N
argument, if enabled, will attempt to disconnect any write sessions currently
directed at your tty.
This is meant to allow users to slam the door on unwelcome writers.
Note that a simple ``mesg n'' will not stop anyone who is already writing
you from continuing to do so, it only prevents new connections from being
made.
The
.B NE
setting also causes a disconnect, but turns your settings to
.B ne
instead of
.BR n .
The
.B d
argument causes a disconnect, just like ``mesg N'', but does not change
your message permissions.
.PP
Normally 
.I mesg
always depermits your tty device, so you can only be written through
.I write
and similar programs.
This prevents arbitrary stuff from being redirected to your tty.
When you do ``mesg Y''
your tty is write permitted to others.
This is rarely necessary or desirable.
.PP
.I Mesg
can also be used to set other switches that affect Orville
.IR write (1).
The
.B -p
flag lets you set preferences to
(w) write, (t) telegrams, (k) talk, or (a) any.
The default is ``any.''
If you set a preference to write,
then people will not be able to send telegrams or talk requests to you.
If they try to send telegrams, they will be asked if they want to write
you instead.
Similarly if you prefer telegrams, people will not be able to write or talk to
you, and if you prefer talk, people will not be able to write or tel you.
You can designate two preferences, like ``mesg -pt -pw'' to allow people to
write or telegram you, but not make talk requests to you.
Alternately, you can use the
.B -x
flag to block particular programs.
Doing ``mesg -xk'' blocks only the talk program, and is equivalent to
``mesg -pt -pw''.
Similarly the ``-xn'' flag excludes no programs and is equivalent to ``-pa''.
Trying to block all programs just turns you permissions off.
.PP
The
.B -m
flag lets you set modes to
(l) line, (c) character, or (a) any.
The default is ``any.''
If you set a mode, then all writes to you will be done in that mode.
If you leave it as ``any,''
the choice is left to the writer.
This will not affect connections already in progress, only future ones.
.PP
The
.B -r
flag lets you turn on or off the recording to telegrams sent to you.
If it is enabled, everytime you are sent a telegram (or a write with
input taken from a file), the text of the messages is saved in a file
named .lastmesg in your home directory.  This allows you to redisplay
the last message sent to you using the
.IR huh (1)
command.
If a screen clear ate a telegram message before you had time to read it,
then the
.I huh
command lets you see it again.
Note that only the last message sent is stored.
The file is permitted to be readable to you only.
.PP
The
.B -b
flag lets you tell the
.I write
and
.I talk
programs whether or not to beep
when a person writes you or sends you a telegram.  The default is to beep.
.PP
The 
.B -h
flag lets you turn on or off your helper status.
People who designate themselves as helpers are announcing their
willingness to help out lost users.
Their accounts will be marked on the
output of the
.IR finger (1)
command, and if anyone does a
.I write
or
.IR jot (1)
to ``help'' they automatically get connected to someone who has a help
flag set.
Normally, turning your permissions off also turns your helper-status off,
but if you set the
.I -h
flag to
.IR Y ,
then you will remain a helper even when your message permissions are off.
This means you can receive help requests, but not normal messages.
.PP
On some systems there is a restricted list of users who may be helpers.
This is usually kept in the file /etc/helpers, one login name per line.
If such a file exists then you will have to get the operators to add your
name to it to be able to designate yourself as a helper.
.PP
If no new settings are given to
.IR mesg ,
then it just reports on the current settings.
Normally it prints the message permissions, but if a
.BR -h ,
.BR -p ,
.BR -r ,
or 
.B  -m
flag was given without a new value after it, then the current status of
that switch will be printed instead.
If you use the
.B -s
flag, then this output will be suppressed.
The command still reports the status of the selected switch with its numeric
return code.
.PP
If you use the
.B -v
flag, all switch settings will be reported in a verbose mode.
.PP
The numeric values returned as return codes
(see below) can also be used to set switches.
Thus ``mesg 0 -m2'' sets permissions on, and the mode to any.
This makes it easy for shell scripts to restore settings that were stored
previously.
.PP
The argument syntax is actually a lot looser than mentioned above.
The dashes before options may be omitted,
Spaces may be added or omitted anywhere in the argument list.

.SH FILES
/dev/tty*
.br
/etc/wrttmp
.br
/etc/helpers
.SH "SEE ALSO"
write(1), amin(1), finger(1), huh(1), helpers(1), talk(1)
.SH DIAGNOSTICS
Exit status is -1 on an error.
Otherwise a code is returned reporting the status of one of the settings.
If the arguments included
.BR -h ,
.BR -p ,
or 
.B  -m
flags without a new value after it, then the last of these listed will be
reported.
Otherwise, if any options were set, the last of those listed in the argument
list will be reported.
And if nothing was set, then message permissions are reported.
.PP
When message permissions, record settings, or helper settings are reported,
0 indicates 'y', and 1 indicates 'n'.
When preferences are reported,
1 indicates 'w', 2 indicates 't', and 4 indicates 'k', and any combinations
are returned as sums of these values.
When modes are reported,
0 indicates 'l', 1 indicates 'c', and 2 indicates 'a'.
.SH BUGS
Turning off 'talk' permissions will only work if you have a talkd which has
been modified to understand Orville write's permission.


CHANGES
 Version 2.55:   Sep 28, 2004
   - Rename log() function to wrtlog() to avoid conflicts with math.h log()
     function.
   - Add distclean target to Makefile.

 Version 2.54:   May 30, 2003
   - Fixed a couple buffer overrun problems that might have constituted a
     possible security hole.
   - Small portability fixes.

 Version 2.53:   Jul 6, 2000
   - Minor portability improvements to install scripts.

 Version 2.52:   Jun 19, 2000
   - Fixed a buffer overflow possibility in huh.c.  This is a major problem if
     'huh' is installed as root.  Older versions should not be installed that
     way.

 Version 2.51:   Feb 20, 2000
   - Renamed lib_utmp.c as getutent.c, and cleaned it up to work more similarly
     to the standard getutent() functions.  Fixed some bugs Karyl Stein pointed
     out with Linux utmp files.

 Version 2.50: Jan 28, 2000
   - Add -pk and -x* flags to mesg, so if we have a cooperative talkd, we can
     control talk permissions seperately from write/tel permissions.
   - Can optionally include hostnames in "Message from" headers.
   - If a user is logged on more than once, write least idle tty.
   - Use autoconf to generate ./configure script which makes Makefile and
     and config.h.
   - Replace old wrttab file with new orville.conf file, and move most site
     configuration into there.
   - Include timezone in message header, since sender's timezone is not
     necessarily the same as the recipients.
   - Lots of rearrangements of header files and such.
   - Miscellaneous protability improvements.
   - Write -r option supported for more different kinds of password files.
   - Changed default name of the 'write -c' link to write from 'chat' to 'jot'.
     Chat is a better name, and I had it first, but the other chat is more
     widely distributed.

 Version 2.41:  Mar 17, 1999
   - ANSIfication of all function declarations.
   - Use getutent() on systems that have it.  Need this when a utmp daemon
     is used instead of a utmp file (thanks to Shane Wegner <shane@cm.nu>).
   - Be a bit more careful about printing/deleting .lastmesg to ensure we
     aren't root when we do it.
   - clean target added to Makefile (thanks to Shane Wegner <shane@cm.nu>).
   - Renamed "common.c" to "lib_common.c" because I wanted to.
   - Lots of minor code cleanup.

 Version 2.40:
   - "mesg -bn" turns off bells.
   - Fixed bug in disconnect so it disconnects multiple writes.
   - exceptions through .yeswrite and .nowrite files and "mesg ye/ne" setting
   - telegram mode no longer sets wrt_what, since it is too transient.
   - telegram mode ignores -p, -y, -n, -c, -l flags (unless switch to write)
   - added helpers command.
   - Day of week in log file dates.
   - mesg command prints warning if people still can write you.
   - If previous helper is still available, another "write help" selects him.
   - Disable core dumps and check that stderr is open.
   - Security patches to .lastmesg file creation. [e]
   - Security patches to .lastmesg reading. [f]
   - Portability improvements to input flushing code

 Version 2.34:
   - Sizes of wrttmp entries stored in wrttmp header.
   - Can write back to someone writing you even if he prefers tels.
   - "mesg d" is like "mesg N" but doesn't change permissions.
   - Linux bug fixes.
   - Made writhist more resistant to system clock errors.

 Version 2.33:
   - Include uid name in message banner if different from utmp name.
   - Root can override recipient's write/tel preferences.
   - Optional recording of telegrams (with mesg -r, amin -p and huh)
   - "mesg -hY" makes you a helper even when permissions are off.

 Version 2.32:
   - wrthist file added, permitting tel replys and regulating flooding.
   - expand control-characters on local echo as well as send.
   - teach it to echo tabs and backspaces more accurately.
   - some low-grade word-wrap in character mode.
   - HELPLOG option for logging help requests only.

 Version 2.31:
   - "write ." writes again to the last user written.
   - write/telegram and line/character mode preferences added.
   - support for login names with more than 8 characters.
   - optional logging of all write requests.
   - added -v to mesg. [2.31e]
   - NOHELPER defines a file to print if there are no helpers. [2.31f]

 Version 2.30:
   - Integration of telegram option into write.
   - Fix bug with shell escapes in "write -n" making "mesg n" perminant.
   - Shell escapes use user's SHELL instead of always using /bin/sh.

 Version 2.20:
   - Reorganization of source code for greater modularity.
   - If multiple helpers are available, choose one at random.
   - When input is redirected, don't ask use stdin to ask user if he wants
     to interupt a conversation.
   - Don't go into character mode till other person has replied.
   - Starting a line with ) in character mode does that line in line mode.
   - If compiled with the NOVICEHELP variable and run with the NOVICE
     environment variable set, prints the contents of the file named in
     NOVICEHELP before running.
   - Fixed amin's job control.

 Version 2.11:
   - If user is logged on more than once, prefer tty that is permitted.
   - Clearer "Permission denied" message.

 Version 2.12:
   - Added TTY_OTHERS

 Version 2.10:
   - Attempted SYSV Compatible system.
   - Added HELPERS

 Version 2.00:  First BSD Compatible system.
   - Added wrttab.
   - Added consistancy checking between utmp and wrttmp to eliminate need for
     modifications to /bin/login.
   - Added support for job control and other BSD-isms.

 Version 1.05:  Last M-Net system III specific version.

 Version 1.00:  First release - Sept 1985.


mesg.c
/*           MESG command for Orville write
 *
 * This 'mesg' command is an extended version of the standard Unix 'mesg'
 * command.  It is designed to work with Orville write
 *
 * (C) Copyright, Sept 1985 - Jan Wolter
 *
 */

#include "orville.h"
#include <signal.h>
#include <pwd.h>
#include <sys/stat.h>

char *mytty;			/* my tty name in tty?? format */
long mypos;			/* offset of my entry in wrttmp file */
struct wrttmp mywrt;		/* my wrttmp entry */
struct utmpx myutmp;		/* A tmp buffer for reading utmp entries */

char silent= FALSE;		/* generates no output if true */
int verbose= FALSE;		/* generate whole table of output if true */

#define SMESG 0
#define SHELP 1
#define SRECO 2
#define SPREF 3
#define SMODE 4
#define SBELL 5
#define SEXCP 6		/* This must be last flag */
#define S_NUM 7

char *name[S_NUM]= {"mesg", "helper", "record", "preference",
                    "mode", "bells", "exceptions"};
char *vname[S_NUM]= {"messages", "helper flag", "record",
#ifdef WITH_TALK
		     "preference (write/tel/talk)",
#else
		     "preference (write/tel)",
#endif
		     "mode (line/character)",
		     "bells", "exceptions"};

/* Legal tag values for each type of setting.  These must be listed in the
 * order of the corresponding numerical codes. nmap lists all legal flags,
 * smap lists only the flags that may be set by "mesg" and must have only
 * truncated versions of the nmap strings. */
char *nmap[S_NUM]= {"yn","ynY","yna","wtkan","lca","yn","yn"};
char *smap[S_NUM]= {"yn","ynY","yn","wtkan","lca","yn","yn"};

char *vmap[S_NUM][4]= { {"yes","no","yes (exceptions)", "no (exceptions)"},
			{"yes","no","always (Y)",NULL},
			{"yes","no","all",NULL},
			{NULL}, /* not used */
			{"line","character","any",NULL},
			{"yes","no",NULL,NULL},
			{"yes","no",NULL,NULL}};

int report= SMESG;		/* which of the above to put in return code */

void set_perms(char new_mesg, int open_up);
int report_status(void);
int get_status(int rep);
void find_me(void);
int may_help(void);
void do_disconnect(void);
void window_warning(int newmode);
char *myhomedir(void);
int maywriteme(char *user, int mode);

int main(int argc, char **argv)
{
int i,j;
char new[S_NUM] = {'x', 'x', 'x', 0x00, 'x', 'x', 'x'};
char disconnect= FALSE;
char open_up = FALSE;
int setting= SMESG;
int flip= 0;
int update;
char wassilent= FALSE;
int lastset= -1;
char flag;

    /* Check if I have a valid tty */
    if (!(fcntl(2,F_GETFL,0) & O_RDWR) || !isatty(2))
    {
	printf("%s: I/O stream 2 improperly redirected\n",argv[0]);
	exit(2);
    }

    /* Find my tty line */
    if (getdevtty()) exit(1);
    mytty= mydevname+5;
    progname= leafname(argv[0]);

    /* Read configuration file */
    readconfig(NULL);

    /* Parse options */
    for(i=1;i<argc;i++)
        for (j= 0; argv[i][j] != '\0'; j++)
	{
	    flag= argv[i][j];

	    /* Convert number arguments to appropriate characters */
	    if (flag >= '0' && flag <= '9')
	    {   if (setting != SPREF)
	    	{
		    if (flag < '0' + strlen(smap[setting]))
			flag= smap[setting][flag-'0'];
		}
		else
		{
		    /* could be multi-digit number - add it up */
		    flag-= '0';
		    while (argv[i][j+1] >= '0' && argv[i][j+1] <= '9')
		    {
			flag= 10*flag + argv[i][++j] - '0';
		    }
		    if (flag > TELPREF_ALL || flag < 0)
		    	goto usage;
		    new[SPREF]= flag;
		    continue;
		}
	    }
	    else
		flag= argv[i][j];

	    switch (flag)
	    {
	    case '-':
		if (j != 0) goto usage;
		break;

	    case 'v':
		verbose= TRUE;
		break;

	    case 's':
		if (setting != SMESG) goto usage;
		wassilent= silent= TRUE;
		break;

	    case 'm':
		if (setting != SMESG) report= setting;
		setting= SMODE;
		break;

	    case 'b':
		if (setting != SMESG) report= setting;
		setting= SBELL;
		break;

	    case 'r':
		if (setting != SMESG) report= setting;
		setting= SRECO;
		break;

	    case 'p':
		if (setting != SMESG) report= setting;
		setting= SPREF;
		flip= 0;
		break;

	    case 'x':
		if (setting != SMESG) report= setting;
		setting= SPREF;
		if (flip == 0) new[SPREF]= TELPREF_ALL;
		flip= 1;
		break;

	    case 'h':
		if (!f_helpers)
		{
		    printf("%s: Cannot use 'h' flag.  Helpers not enabled "
		    	"on this system.\n", progname);
		    exit(1);
		}
		if (setting != SMESG) report= setting;
		setting= SHELP;
		break;

	    case 'E':
	    case 'e':
		if (!f_exceptions)
		{
		    printf("%s: Cannot use '%c' flag.  Exceptions not enabled "
		    	"on this system.\n", progname, flag);
		    exit(1);
		}
		if (lastset != SMESG) goto usage;
		new[SEXCP]= 'y';
		break;

	    case 'Y':
		if (setting == SMESG)
		{
		    open_up= TRUE;
		    new[SMESG]= 'y';
		    lastset= SMESG;
		    break;
		}
		else if (setting != SHELP)
		    goto usage;
		/* else drop through */

#ifdef WITH_TALK
	    case 'k':
#endif
	    case 'y':
	    case 'n':
	    case 'l':
	    case 'c':
	    case 'w':
	    case 't':
	    case 'a':
		if (strchr(smap[setting],flag) == NULL) goto usage;
		if (setting == SPREF)
		{
		    if (flip)
		    {
			switch (flag)
			{
			case 'w': new[SPREF]&= ~TELPREF_WRITE; break;
			case 't': new[SPREF]&= ~TELPREF_TEL; break;
#ifdef WITH_TALK
			case 'k': new[SPREF]&= ~TELPREF_TALK; break;
#endif
			case 'n': new[SPREF]= TELPREF_ALL; break;
			case 'a': goto usage;
			}
		    }
		    else
		    {
			switch (flag)
			{
			case 'w': new[SPREF]|= TELPREF_WRITE; break;
			case 't': new[SPREF]|= TELPREF_TEL; break;
#ifdef WITH_TALK
			case 'k': new[SPREF]|= TELPREF_TALK; break;
#endif
			case 'a': new[SPREF]= TELPREF_ALL; break;
			case 'n': goto usage;
			}
		    }
		}
		else
		    new[setting]= flag;
		lastset= setting;
		setting= SMESG;
		break;

	    case 'N':
		lastset= SMESG;
	    case 'd':
		if (setting != SMESG) goto usage;
		disconnect= TRUE;
		if (flag == 'N') new[SMESG]= 'n';
		break;

#ifndef WITH_TALK
	    case 'k':
	    	printf("%s: Talk permissions cannot be separately set "
		    "at this site\n",progname);
		goto usage;
#endif

	    default:
		goto usage;
	    }
	}

    /* If we changed message perms, but didn't set exceptions, turn them off */
    if (new[SMESG] != 'x' && new[SEXCP] == 'x') new[SEXCP]= 'n';

    if (setting != SMESG)
	report= setting;
    else if (report == SMESG)
	report= ((lastset > SMESG) ? lastset : SMESG);

    /* Open the wrttmp file and find us in it */
    if (init_wstream(O_RDWR))
	exit(-1);
    find_me();

    if ((new[SHELP] == 'y'  || new[SHELP] == 'Y') && !may_help())
    {
	printf("You are not on the helper list.\n");
	exit(1);
    }

    /* Update permissions/helperness/preferences/bells */
    update= 0;
    if (new[SMESG] != 'x')
    {	set_perms(new[SMESG],open_up); update= 1; }
    if (f_helpers && new[SHELP] != 'x')
    {	mywrt.wrt_help= new[SHELP]; update= 1; }
    if (f_exceptions && new[SEXCP] != 'x')
    {	mywrt.wrt_except= new[SEXCP]; update= 1; }
    if (new[SRECO] != 'x')
    {	mywrt.wrt_record= new[SRECO]; update= 1; }
    if (new[SMODE] != 'x')
    {	mywrt.wrt_modepref= new[SMODE]; update= 1; }
    if (new[SPREF] != 0x00 && !flip)
    {	mywrt.wrt_telpref= new[SPREF]; update= 1; }
    if (new[SPREF] != TELPREF_ALL && flip)
    {	if (new[SPREF] == 0x00)
            set_perms('n',0);
        else
	    mywrt.wrt_telpref= new[SPREF];
	update= 1;
    }
    if (new[SBELL] != 'x')
    {	mywrt.wrt_bells= new[SBELL]; update= 1; }

    /* Write out new settings */
    if (update)
    {
	lseek(wstream,mypos,0);
	write(wstream,&mywrt,sizeof(struct wrttmp));
	silent= TRUE;
    }

    /* Disconnect existing connections */
    if (disconnect)
    {
    	if (f_disconnect)
	    do_disconnect();
	else
	    printf("Disconnect option not enabled on this system\n");
        silent= TRUE;
    }

    /* Close the utmp file */
    endutxent();

   if (f_wrthist != NULL &&
       !wassilent && (new[SMESG] == 'n' || new[SEXCP] == 'y'))
       window_warning((new[SMESG]=='n')+2*(new[SEXCP]=='y'));

    exit(report_status());

usage:
    printf("usage: %s [-sv] [y|n", progname);
    if (f_disconnect) printf("|d");
    if (f_exceptions) printf("|ye|ne");
    printf("|Y");
    if (f_disconnect)
    {
    	printf("|N");
	if (f_exceptions)
	    printf("|NE");
    }
    printf("] ");
    if (f_helpers) printf(" [-h[y|n|Y]]");
    printf(" [-r[y|n]] [-m[l|c|a]] [-p[w|t|k|a]] [-x[w|t|k|n]] [-b[y|n]]\n");
    exit(-1);
}


/* SET_PERMS:  This sets the message permissions to "new_mesg" (which is
 * either 'y' or 'n'.  It throws the tty permissions wide open as well if
 * "open_up" is true.
 */

void set_perms(char new_mesg, int open_up)
{
int mode;

    /* Change the device permissions */
    if (open_up)
	mode= PERMS_OPEN;
#ifdef TTYPERMS
    else if (new_mesg == 'y')
	mode= PERMS_ON;
#endif /*TTYPERMS*/
    else
	mode= PERMS_OFF;

    chmod(mydevname,mode);

    /* Change the wrttmp permissions */
#ifndef TTYPERMS
    mywrt.wrt_mesg= new_mesg;
#endif /*TTYPERMS*/
}


/* REPORT_PREF -- This prints a message reporting the perf setting, in
 * verbose or non-verbose format.  The appropriate numeric return code is
 * returned.
 */

int report_pref(int verbose)
{
int val,n;

    if (!silent)
    {
	if (verbose)
	    printf("%s: ",vname[SPREF]);
	else
	    printf("%s is ",name[SPREF]);
    }

    val= mywrt.wrt_telpref;
    
    if (val & TELPREF_OLD)
    {
    	switch (val)
	{
	case 'w': val= TELPREF_WRITE; break;
	case 't': val= TELPREF_TEL; break;
#ifdef WITH_TALK
	case 'k': val= TELPREF_TALK; break;
#endif
	case '-': case 'a': val= TELPREF_ALL; break;
	}
    }

    if (!silent)
    {
	if (val == TELPREF_ALL)
	{
	    fputs(verbose ? "all\n":"a\n" ,stdout);
	    return val;
	}
	n= 0;
	if (val & TELPREF_WRITE)
	{
	    fputs(verbose ? "write":"w" ,stdout);
	    n++;
	}
	if (val & TELPREF_TEL)
	{
	    if (n > 0) putchar(',');
	    fputs(verbose ? "tel":"t" ,stdout);
	    n++;
	}
#ifdef WITH_TALK
	if (val & TELPREF_TALK)
	{
	    if (n > 0) putchar(',');
	    fputs(verbose ? "talk":"k" ,stdout);
	}
#endif
	putchar('\n');
    }
    return val;
}

/* REPORT_STATUS -- This prints the current message permission settings and
 * returns the status code (0 = yes; 1 = no).  If silent is true, it omits
 * the printing.  If report is set to something other than SMESG it reports
 * on that other setting.  If verbose is true, it reports on all settings.
 */

int report_status()
{
int code,rc,i;

    if (verbose)
    {
	for (i= 0; i < S_NUM; i++)
	{
		if (i == SEXCP) continue;
		if (!f_helpers && i == SHELP) continue;
		if (i == SPREF) {report_pref(1); continue; }
		code= get_status(i);
		printf("%s: %s\n",vname[i],
		    vmap[i][code] != NULL ? vmap[i][code] : "strange");
		if (i == report) rc= code;
	}
    }
    else
    {
	rc= get_status(report);
	if (!silent)
	{
	    if (f_exceptions && report == SMESG)
		printf("%s is %c%s\n",name[report],nmap[report][rc%2],
		       rc > 1 ? "e" : "");
	    else if (report == SPREF)
	        report_pref(0);
	    else
		printf("%s is %c\n",name[report],nmap[report][rc]);
    	}
    }
    return rc;
}


/* GET_STATUS -- Get current index code of given switch */

int get_status(int rep)
{
struct stat stb;
int code;

    switch (rep)
    {
    case SMESG:
#ifdef TTYPERMS
	/* Stat my tty */
	if (stat(mydevname,&stb))
	{
	    printf("%s: Error - can't stat %s\n",progname,mydevname);
	    exit(-1);
	}
	code= ((stb.st_mode & 0020)==0);
#else
	code= strchr(nmap[SMESG],mywrt.wrt_mesg) - nmap[SMESG];
#endif /* TTYPERMS */
	if (f_exceptions)
	    code+= 2 - 2*(strchr(nmap[SEXCP],mywrt.wrt_except) - nmap[SEXCP]);
	break;

    case SEXCP:
	code= strchr(nmap[SEXCP],mywrt.wrt_except) - nmap[SEXCP];
	break;

    case SHELP:
	code= strchr(nmap[SHELP],mywrt.wrt_help) - nmap[SHELP];
	break;

    case SRECO:
	code= strchr(nmap[SRECO],mywrt.wrt_record) - nmap[SRECO];
	break;

    case SMODE:
	code= strchr(nmap[SMODE],mywrt.wrt_modepref) - nmap[SMODE];
	break;

    case SBELL:
	code= strchr(nmap[SBELL],mywrt.wrt_bells) - nmap[SBELL];
	break;
    }
    return code;
}


/* DO_DISCONNECT -- This does the disconnect, by searching the wrttmp file for
 * people who are writing you and sending a SIGTERM to each of the process
 * numbers listed there.  This whole approach is a bit shady.
 */

void do_disconnect()
{
struct utmpx *ut;		/* A tmp buffer for reading utmp entries */
struct wrttmp hiswrt;		/* Someone's wrttmp entry */
int slot= 0;

    /* Rewind utmp file */
    setutxent();

    /* For each user who is writing me */
    for (;;)
    {
	lseek(wstream, wrttmp_offset(slot++), 0);
	if (read(wstream, &hiswrt, sizeof(struct wrttmp)) !=
	       sizeof(struct wrttmp))
	   break;

	if (!strncmp(hiswrt.wrt_what, myutmp.ut_user, sizeof(myutmp.ut_user)))
	{
	    setutent();
	    /* Check apparant writer against utmp file */
	    while ((ut= getutxent()) != NULL)
		if (
#ifdef USER_PROCESS
		    ut->ut_type == USER_PROCESS &&
#endif
		    !strncmp(hiswrt.wrt_line, ut->ut_line, sizeof(ut->ut_line)))
		{
		    /* Writer is for real: bonk him one */
		    kill(hiswrt.wrt_pid, SIGTERM);
		    break;
		}
	}
    }
}



/* FIND_ME -- This finds my wrttmp entry and loads it into "mywrt".
 */

void find_me(void)
{
struct utmpx *ut;

    /* Find our entry in the Utmp file */
    if ((ut= find_utmp(mytty)) == NULL || ut->ut_user[0] == '\0')
    {
	printf("%s: Unable to find your tty (%s) in utmp file\n",
		progname,mytty);
	exit(-1);
    }
    myutmp= *ut;

    /* Find the entry in the wrttmp file */
    find_wrttmp(mytty,myutmp.ut_tv.tv_sec,&mywrt,&mypos);
}


/* MAY_HELP -- Check if the user's name is anywhere to be found in the
 * f_helperlist file.  If he is there, or the file doesn't exist at all, then
 # he may designate himself as a helper.
 */

int may_help()
{
#define BUFSZ 80
FILE *hfp;
char buf[BUFSZ+1];
char myname[sizeof(myutmp.ut_user)+1];

	if (f_helperlist == NULL || (hfp= fopen(f_helperlist,"r")) == NULL)
		return TRUE;
	
	strncpy(myname,myutmp.ut_user,sizeof(myutmp.ut_user) -1);
	myname[sizeof(myutmp.ut_user)]= '\0';
	strcat(myname,"\n");
	
	while (fgets(buf,BUFSZ,hfp) != NULL)
		if (!strcmp(buf,myname))
		{
			fclose(hfp);
			return TRUE;
		}

	fclose(hfp);
	return FALSE;
}


/* WINDOW_WARNING - If there are people who can still write you under the
 * four minute window rule, print a warning message.
 */

void window_warning(int newmode)
{
struct wrthist *hist;
struct wrttmp w;
struct utmpx *u;
long writer, writee;
time_t now;
int n, foundsome= 0;
FILE *fp;

    if (newmode == 0) return;

    /* Compute my slot number from my offset */
    if ((writer= (mypos - wt_head.hdr_size)/wt_head.tmp_size) > MAXSLOTS)
	return;

    if ((fp= fopen(f_wrthist,"r")) == NULL)
	return;

    fseek(fp,whindex(writer,0),0);
    hist= (struct wrthist *)malloc(MAXSLOTS*sizeof(struct wrthist));
    n= fread(hist,sizeof(struct wrthist),MAXSLOTS,fp);
    fclose(fp);

    now= time((time_t *)0);

    for (writee= 0; writee < n; writee++)
    {
	if (hist[writee].tm > myutmp.ut_tv.tv_sec &&
	    now - hist[writee].tm <= f_answertel)
	{
	    /* Fetch "his" wrttmp entry - it may actually belong to a previous
	     * user on that line, so only the wrt_line field is really to be
	     * trusted.
	     */
	    lseek(wstream,wrttmp_offset(writee),0);
	    if (read(wstream, &w, sizeof(struct wrttmp)) !=
		   sizeof(struct wrttmp))
		continue;

	    /* Check times to ensure this is not from a later user */
	    if (hist[writee].tm < w.wrt_time)
	    	continue;

	    /* Fetch his utmp entry, and confirm that the current user was
	     * already logged in there when we sent our last telegram there.
	     */
	    if ((u= find_utmp(w.wrt_line)) == NULL || u->ut_user[0] == '\0' ||
	        hist[writee].tm < u->ut_tv.tv_sec)
	    	continue;

	    /* Check if due to exceptions he may write us anyway */
	    if (f_exceptions && newmode > 1 && maywriteme(u->ut_user, newmode))
	    	continue;

	    if (!foundsome)
	    {
	        printf("Warning: Users to whom you recently sent messages may\n");
	        printf("still be able to write you for a few more minutes:\n");
	        foundsome= 1;
	    }
            int usize = (int)sizeof(u->ut_user) - 1;
            int lsize = (int)sizeof(u->ut_line) - 1;
	    printf("  %-*.*s %-*.*s %4.1f more minutes\n",
	    	usize, usize, u->ut_user, lsize, lsize, u->ut_line,
	        (float)(f_answertel - now + hist[writee].tm)/60.0);
	}
    }
}


/* MYHOMEDIR - Return my home directory.  Speed counts over reliability here.
 * Return NULL if we can't find it.
 */

char *myhomedir()
{
char myname[sizeof(myutmp.ut_user)+1];
struct passwd *pw;
char *dir;

    /* Try environment variable first */
    if ((dir= getenv("HOME")) != NULL)
	return dir;

    /* If that don't work, try passwd file */
    strncpy(myname,myutmp.ut_user,sizeof(myutmp.ut_user) -1);
    myname[sizeof(myutmp.ut_user)]= '\0';
    if ((pw= getpwnam(myname)) != NULL)
    	return pw->pw_dir;

    /* If that don't work, give it up */
    return NULL;
}


/* MAYWRITEME - can the named user write me?  This checks exceptions.  Does
 * some caching so that multiple successive calls are efficient.
 */

int maywriteme(char *user, int mode)
{
static int oldmode= -1;
static char *list;
struct stat st;
char *dir,*fname, *p,*q;
FILE *fp;
int n, l, nl, sp;

    if (mode < 2)
	return (mode == 0);

    if (mode != oldmode)
    {
	list= NULL;
	oldmode= mode;

	/* Open the appropriate exception file */
	if ((dir= myhomedir()) == NULL)
	    return FALSE;
	fname= (char *)malloc(strlen(dir)+15);
	sprintf(fname,"%s/%s",dir, mode==2 ? ".nowrite" : ".yeswrite");
	fp= fopen(fname,"r");
	free(fname);

	/* Load the file into memory */
	if (fp != NULL && !fstat(fileno(fp),&st))
	{
	    list= (char *)malloc(st.st_size+1);
	    n= fread(list,1,st.st_size,fp);
	    list[n]= '\0';
	}

	if (list != NULL)
	{
	    /* Compress list into login names separated by single spaces */
	    nl= 1; sp= 1;
	    for (p= q= list; *p != '\0'; p++)
	    {
		/* Skip lines starting with # */
		if (nl && (*p == '#'))
		{
		    p= strchr(p,'\n');
		    if (p == NULL) break;
		}

		nl= (*p == '\n');

		if (strchr(" ,;:\t\n\r",*p))
		{
		    if (!sp) *(q++)= ' ';
		    sp= 1;
		}
		else
		{
		    sp= 0;
		    *(q++)= *p;
		}
	    }
	    *q= '\0';
	}
    }

    if (list != NULL)
    {
	/* Find length of user name */
	for (l= 0; l < UT_NAMESIZE && user[l] != '\0'; l++)
	    ;

	/* Search compressed list for the name */
	p= list;
	for (;;)
	{
	    if (!strncmp(p,user,l) && (p[l] == ' ' || p[l] == '\0'))
	    	return (mode != 2);
	    if ((p= strchr(p,' ')) == NULL) break;
	    p++;
	}
    }
   
    /* Not on list */
    return (mode == 2);
}
